src/bin/pg_autoctl/pgsetup.c

```diff
@@ -1122,3 +1134,199 @@ pgsetup_get_pgport()
+	 */
+	if (ssl->active && !ssl->createSelfSignedCert)
+	{
+		/* we say allFilesGiven but we can do without the CRL file */
+		bool allFilesGiven = !IS_EMPTY_STRING_BUFFER(ssl->caFile)
+			&& !IS_EMPTY_STRING_BUFFER(ssl->serverCert)
+			&& !IS_EMPTY_STRING_BUFFER(ssl->serverKey);
+
+		if (!allFilesGiven)
+		{
+			log_error("Failed to setup SSL with user-provided certificates: "
+					  "options --ssl-ca-file --ssl-server-cert --ssl-server-key"
+					  "are required.");
+			return false;
+		}
+
+		/* when given all the files, check they exist */
+		if (!file_exists(ssl->caFile))
+		{
+			log_error("--ssl-ca-file file does not exist at \"%s\"",
+					  ssl->caFile);
+			return false;
+		}
+
+		if (!IS_EMPTY_STRING_BUFFER(ssl->crlFile) && !file_exists(ssl->crlFile))
+		{
+			log_error("--ssl-crl-file file does not exist at \"%s\"",
+					  ssl->crlFile);
+			return false;
+		}
+
+		if (!file_exists(ssl->serverCert))
+		{
+			log_error("--server-crt file does not exist at \"%s\"",
+					  ssl->serverCert);
+			return false;
+		}
+
+		if (!file_exists(ssl->serverKey))
+		{
+			log_error("--server-key file does not exist at \"%s\"",
+					  ssl->serverKey);
+			return false;
+		}
+
+		/* install a default value for --ssl-mode, use verify-full */
+		if (ssl->sslMode == SSL_MODE_UNKNOWN)
+		{
+			ssl->sslMode = SSL_MODE_VERIFY_FULL;
+			strlcpy(ssl->sslModeStr,
+					pgsetup_sslmode_to_string(ssl->sslMode), SSL_MODE_STRLEN);
+			log_info("Using default --ssl-mode \"%s\"", ssl->sslModeStr);
+		}
+
+		return true;
+	}
+
+	/*
+	 * When --ssl-self-signed is used, we default to using sslmode=require.
+	 * Setting higher than that are wrong, false sense of security.
+	 */
+	if (ssl->createSelfSignedCert)
+	{
+		/* in that case we want an sslMode of prefer at most */
+		if (ssl->sslMode > SSL_MODE_PREFER)
+		{
+			log_error("--ssl-mode \"%s\" is not compatible with self-signed "
+					  "certificates, please provide certificates signed by "
+					  "your trusted CA.",
+					  pgsetup_sslmode_to_string(ssl->sslMode));
+			log_info("See https://www.postgresql.org/docs/current/libpq-ssl.html"
+					 " for details");
+			return false;
+		}
+
+		if (ssl->sslMode == SSL_MODE_UNKNOWN)
+		{
+			/* install a default value for --ssl-mode */
+			ssl->sslMode = SSL_MODE_REQUIRE;
+			strlcpy(ssl->sslModeStr,
+					pgsetup_sslmode_to_string(ssl->sslMode), SSL_MODE_STRLEN);
+			log_info("Using default --ssl-mode \"%s\"", ssl->sslModeStr);
+		}
+
+		log_info("Using --ssl without certificates: pg_autoctl will "
+				 " create self-signed certificates, allowing for "
+				 "encrypted network traffic");
+		log_warn("Self-signed certificates provide protection against "
+				 "eavesdropping; this setup does NOT protect against "
+				 "Man-In-The-Middle attacks nor Impersonation attacks.");
+		log_warn("See https://www.postgresql.org/docs/current/libpq-ssl.html "
+				 "for details");
+
+		return true;
+	}
+
+	/* --no-ssl is ok */
+	if (ssl->active == 0)
+	{
+		return true;
+	}
+
```

Comment:
There should be an error here that it is required to specify an ssl flag

Reply:
Good spot! It's not where to implement the error catching, because we're lacking the information at this place, so I made it happen in the main argument parsing function... which means in two places, unfortunately.